path-util: introduce path_simplify()

yuwata opened the pull request:
This slightly relaxes the restriction for specifying paths in unit files. Previously, we do not allow path which starts with `./`, ends with `/.`, or has `/./`. But these are redundant but safe. So, this introduces the function `path_kill_slashes_and_dot()` which cleanups such redundant path specifications.

Fixes #9107.

@poettering You tagged #9107 'needs-better-log-message'. But I think `/./` or trailing `/.` are safe. What do you think about this?

yuwata commented:
@keszybz I noticed right now that you self-assigned the issue. I do not want to disturb your work, but if you are interested in this, please give me a comment. Thank you.

keszybz commented:
I wanted to work on it, but I started doing something else in the mean time. I'll reassign it to you.

yuwata commented:
Updated. PTAL.

keszybz commented on src/test/test-path-util.c:
```diff
@@ -69,15 +78,28 @@ static void test_path(void) {
+        test_path_kill_slashes_one("aaa/bbb////ccc", "aaa/bbb/ccc", "aaa/bbb/ccc");
+        test_path_kill_slashes_one("//aaa/.////ccc", "/aaa/./ccc", "/aaa/ccc");
+        test_path_kill_slashes_one("///", "/", "/");
+        test_path_kill_slashes_one("///.//", "/.", "/");
```
Hmm, this removes a trailing slash, and a trailing slash is significant in some contexts. I'm not sure if this matters here.

yuwata replied:
trailing slash? I think I do not change anything about slashes.

keszybz commented on src/basic/path-util.c:
```diff
@@ -992,3 +993,52 @@ bool empty_or_root(const char *root) {
+
+                if (!absolute && (flag & PATH_CHECK_ABSOLUTE)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0,
+                                   "%s= path is not an absolute%s: %s",
```
"is not absolute"

keszybz commented on src/basic/path-util.c:
```diff
@@ -992,3 +993,52 @@ bool empty_or_root(const char *root) {
+
+                if (absolute && (flag & PATH_CHECK_RELATIVE)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0,
+                                   "%s= path is an absolute%s: %s",
```
"is absolute"

keszybz commented on src/basic/path-util.c:
```diff
@@ -992,3 +993,52 @@ bool empty_or_root(const char *root) {
+                return -EINVAL;
+        }
+
+        if (!utf8_is_valid(path)) {
```
So this check must be first, because otherwise we can print non-utf8 in the messages above.

yuwata replied:
OK. it is moved to the beginning of the function.

keszybz commented on src/core/load-fragment.c:
```diff
@@ -2599,17 +2572,16 @@ int config_parse_unit_requires_mounts_for(
+                r = path_simplify_and_warn(resolved, PATH_CHECK_ABSOLUTE, unit, filename, line, lvalue);
+                if (r < 0)
+                        continue;
+
```
The check for utf8-validity needs to be before `unit_full_printf` and its error message. I don't think you can remove the utf8 check.

yuwata replied:
I see... Are there any possibility that `unit_{full,name}_printf()` generate utf8 invalid strings? I mean, that should we check utf8-validity before _and_ after `unit_{full,name}_printf()`?

yuwata replied:
I added utf8-validity check in `parse_line()`. So, conf parsers do not need to call `utf8_is_valid()` now.

keszybz replied:
That's a great solution.

poettering commented:
Hmm, I haven't liked the name "path_kill_slashes()" since a longer time, I was always attempted to rename it to path_normalize(), to match path_is_normalized(), but then again the latter doesn't permit embedded "..", which we can't really remove in this function as doing that requires hitting the disk... 

That said, maybe a more suitable name would be "path_transform()" or "path_tidy()" or so? @keszybz ideas?

poettering commented on src/test/test-path-util.c:
```diff
@@ -26,6 +26,15 @@
+                assert_se(streq(path_kill_slashes_and_dot(p, false), out)); \
+                p = strdupa(in);                                        \
+                assert_se(streq(path_kill_slashes_and_dot(p, true), out_dot)); \
+        }
```
why make this a macro? why not a regular function? (some test cases use macros like this for no real reason i can see, but we should not add more...)

yuwata replied:
Now I use function instead of macro.

poettering commented:
So, i guess accepting/normalizing inner dots for WorkingDirectory= is OK, given that the field usually references foreign, externally defined paths, hence conceptually the PR looks mostly good. I am wondering howerver if we really should do that for RuntimeDirectory() too.. I mean, the runtime directory is not really a path, it's a part of the of the path, and we build the actual path from it. Hence I think we should be stricter there, and not allow non-normalized stuff... That said I am not feeling too strongly about this, I'd be fine with this PR either way...

poettering commented:
(But I'd still like to see path_kill_slashes_and_dot()/path_kill_slashes() to be renamed to path_simplify() a part of this PR)

poettering commented:
Hmm, the original bug report complained that "systemctl status" for an offending service looks like this:

```
systemctl status bugdemo.service
● bugdemo.service - demonstrate possible bug
Loaded: error (Reason: Operation not permitted)
```

This PR doesn't really change that, if the path contains "..", right? My original plan was to simply make sure that the error we report there is a bit more useful, though not entirely sure how much work that would actually be... property_get_load_error() generates that error string server side, so we could fix this without the client needing changes, but I am not sure how best to fill it with indication that WorkingDirectory= is at fault here... 

Maybe a workable way out is to just generically say that configuration settings are at fault, which should be actionable enough and the logs should then show details.

Lemme cook up a patch for that part. I think it would be orthogonal to your work however, and both should go in

yuwata commented:
Updated. But not addressed all comments. Still wip.

yuwata commented:
Updated. PTAL.

@poettering 
> I am wondering howerver if we really should do that for RuntimeDirectory() too.. I mean, the runtime directory is not really a path, it's a part of the of the path, and we build the actual path from it.

Yes, so I added the flag `PATH_CHECK_RELATIVE`.

> Hence I think we should be stricter there, and not allow non-normalized stuff...

Hmm, sorry, I cannot understand why...
